allow-custom-protocol-navigation

ArthurSonzogni opened the pull request:
For security reasons, web browsers started to block navigations toward
custom protocols initiated from within a sandboxed iframe. They are
allowed, only when normal navigations are already allowed in top-level
frame and when opening new popups.
See: https://github.com/whatwg/html/pull/7124
And: https://github.com/whatwg/html/issues/2191#issuecomment-1006573881

So far so good, but it is hard for some website like
https://teams.microsoft.com/ to handle this change. They would like not
to block custom scheme, while continuing to block supported ones.

An escape hatch could be to mint a new
`allow-custom-protocol-navigation` sandbox flags. This would allow web
browser to ship the initial security mitigation without breaking them.

Bug: https://github.com/whatwg/html/issues/7468

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->

- [ ] At least two implementers are interested (and none opposed):
   * Chrome is: https://crbug.com/1253379
   * Firefox: I will ask on https://github.com/mozilla/standards-positions/issues/581
   * Webkit: I will ask.
- [ ] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * [Similar to the previous PR](https://github.com/whatwg/html/pull/7124), it is not possible to provide WPT tests, because this is at the boundary of the web platform, and can't be observed from within.
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://crbug.com/1253379
   * Firefox: WIP. I think I will update https://bugzilla.mozilla.org/show_bug.cgi?id=1735746 directly.
   * Safari: WIP: I will open one.
   * Deno: N/A
   * Node.js: N/A

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

cdumez commented:
Sounds good to me. WebKit had to introduce a quirk to keep Microsoft Teams working.

domenic commented on source:
```diff
@@ -31263,7 +31264,8 @@ interface <dfn interface>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
   browsing context's <span>active window</span> has <span>transient activation</span>; and the <code
-  data-x="attr-iframe-sandbox-allow-forms">allow-forms</code>, <code
+  data-x="attr-iframe-sandbox-allow-custom-protocols-navigation">allow-custom-protocols-navigation</code>,
+  <code data-x="attr-iframe-sandbox-allow-forms">allow-forms</code>, <code
```
This breaks the "respectively" later in the sentence. Maybe it should get its own clause similar to "allow-top-navigation".

domenic reviewed (changes requested):
I think you need to update the underlying model, at https://html.spec.whatwg.org/#sandboxing . How this section of the spec works is that the keywords get parsed and mapped to a flag set, with some translation, in https://html.spec.whatwg.org/#parse-a-sandboxing-directive:

- Flags prevent things, whereas keywords allow things. So the flags get set, *unless* the keywords are present.

- Some keywords are redundant or combine to create the underlying flags. E.g. both `allow-top-navigation` and `allow-top-navigation-by-user-activation` cause the [sandboxed top-level navigation with user activation browsing context flag](https://html.spec.whatwg.org/#sandboxed-top-level-navigation-with-user-activation-browsing-context-flag) to be removed.

So I think you'll want some sort of "sandboxed custom protocol handlers" flag, which gets removed if any of `allow-top-navigation` or `allow-popups` or `allow-custom-protocols-navigation` are set.

The fact that "hand-off to external software" uses token keywords, instead of flag names, seems like a preexisting spec bug. The token keywords do not exist at this point, and are not the type of thing which a sandboxing flag set contains.

I would offer to fix this so that you could build on a better foundation, but since you are going to change the logic here somewhat dramatically, I think it's best if you fix it as part of the same PR.

ArthurSonzogni commented:
> The fact that "hand-off to external software" uses token keywords, instead of flag names, seems like a preexisting spec bug. The token keywords do not exist at this point, and are not the type of thing which a sandboxing flag set contains.
> [...]
> I would offer to fix this so that you could build on a better foundation,

I see. Thanks!

I was quite happy this PR to be straightforward. it would make me sad to see it polluted by fixing pre existing issues. I would prefer fixing pre existing issues separately. I will prepare another PR to address this.

ArthurSonzogni commented on source:
```diff
@@ -88466,6 +88468,10 @@ interface <dfn interface>Location</dfn> { // but see also <a href="#the-location
     of:</p>
```
Okay.
Since I wasn't able to specify the consequences of using both, I put them into a note:
![image](https://user-images.githubusercontent.com/4759106/155506561-5f5dad32-c005-4d1f-bafc-c309d07ef3a3.png)

Does this sounds good to you?

ArthurSonzogni commented:
Thanks @domenic 
I think I addressed all your comments.

I also turned list of token into `<ul>`, sorted alphabetically. This is more readable. This also make me realized two token were missing in one list. I fixed it.

Some images of the change:

![image](https://user-images.githubusercontent.com/4759106/155508618-a024962c-d86e-4f48-8b7f-f52bc8afe538.png)

---

![image](https://user-images.githubusercontent.com/4759106/155508351-9fac4a41-b0aa-4b6b-92d4-4546a1c0750f.png)

---

![image](https://user-images.githubusercontent.com/4759106/155508231-dbe7ba1a-0cd8-4254-948e-0992d0f238fa.png)

---

![image](https://user-images.githubusercontent.com/4759106/155508865-b02abfbb-7897-4629-8dc1-f2f4a4baf53f.png)

domenic commented on source:
```diff
@@ -31262,18 +31268,20 @@ interface <dfn interface>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
+  browsing context's <span>active window</span> has <span>transient activation</span>; the <code
+  data-x="attr-iframe-sandbox-allow-custom-protocols-navigation">allow-custom-protocols-navigation</code>
+  re-enable navigations toward URLs unsupported by the user agent to be <span>hand-off to external
+  software</span>; and the <code data-x="attr-iframe-sandbox-allow-forms">allow-forms</code>, <code
```
Grammar: I think this should be something like "the allow-custom-protocols-navigation keyword reenables navigations toward URLs unsupported by the user agent to be handed off to external software"

ArthurSonzogni replied:
Done.

domenic commented on source:
```diff
@@ -31284,6 +31292,12 @@ interface <dfn interface>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
   data-x="attr-iframe-sandbox-allow-top-navigation">allow-top-navigation</code> will have an effect
   in such non-conformant markup.</p>
 
+  <p class="note">The <code
```
This should not be a non-normative note; although it doesn't have an impact for browsers, it is still a normative requirement on authors. Phrasing it similar to the previous paragraph should work. E.g.

> Similarly, the [allow-custom-protocols-navigation](https://whatpr.org/html/7654/origin.html#attr-iframe-sandbox-allow-custom-protocols-navigation) keyword must not be specified if either the [allow-top-navigation](https://whatpr.org/html/7654/origin.html#attr-iframe-sandbox-allow-top-navigation) or [allow-popups](https://whatpr.org/html/7654/origin.html#attr-iframe-sandbox-allow-popups) keywords are specified, as doing so is redundant.

ArthurSonzogni replied:
Done.

domenic commented on source:
```diff
@@ -83825,6 +83839,13 @@ interface <dfn interface>BarProp</dfn> {
+   <dt>The <dfn export>sandboxed custom protocols navigation browsing context flag</dfn></dt>
+
+   <dd>
+    <p>This flag prevents navigations toward URLs unsupported by the user agent to be <span>hand-off
```
Grammar again: "prevents navigations toward URLs unsupported by the user agent from being handed off to external software".

ArthurSonzogni replied:
Done.

domenic commented on source:
```diff
@@ -88462,26 +88489,26 @@ interface <dfn interface>Location</dfn> { // but see also <a href="#the-location
-     token and <var>hasTransientActivation</var> is true</p></li>
+    <p>Return without invoking the external sofware package if all of these conditions hold:</p>
+    <ul class="brief">
+      <li><p><var>browsingContext</var> is not a <span>top-level browsing context</span></p></li>
```
Nit: single-space indents.

domenic commented on source:
```diff
@@ -88462,26 +88489,26 @@ interface <dfn interface>Location</dfn> { // but see also <a href="#the-location
-    new popup or a new top-level navigation. Therefore, we enforcing sandbox flags
-    appropriately.</p>
+    new popup or a new top-level navigation. That's why its is allowed in sandboxed iframe only when
+    one of "<code data-x="attr-iframe-sandbox-allow-popups">allow-popups</code>", "<code
```
Let's not add quotes around the keyword names here.

ArthurSonzogni replied:
Done.

domenic commented on source:
```diff
@@ -124458,16 +124485,25 @@ interface <dfn interface>External</dfn> {
-          "<code data-x="attr-iframe-sandbox-allow-scripts">allow-scripts</code>" and
-          "<code data-x="attr-iframe-sandbox-allow-top-navigation">allow-top-navigation</code>"
+      <ul class="brief">
+        <li><code
```
Nit: single-space indent

ArthurSonzogni replied:
Done.

domenic reviewed (commented):
Looks great, just nits and grammar suggestions!

annevk commented:
Should this be `allow-custom-protocols-top-navigation`? (In an ideal world "top" is redundant, but I'm not sure we're quite there yet with `registerProtocolHandler()`.)

ArthurSonzogni commented:
> Should this be `allow-custom-protocols-top-navigation`? (In an ideal world "top" is redundant, but I'm not sure we're quite there yet with `registerProtocolHandler()`.)

Happy to edit this PR if you, or someone else, think this is better. I was slightly reluctant, because this is handed off externally, so this is not really a main frame nor a subframe navigation.

If we do, can we start with the `allow-top-navigation-` prefix? This way, it would share a common prefix with two other sandbox flags. Maybe:
- `allow-top-navigation-to-custom-scheme`
- `allow-top-navigation-to-external-application`

annevk commented:
Well, it's not always handed off externally, is it? For instance, when `registerProtocolHandler()` is used. Ideally they all end up using the same code path though, whether it's translated to HTTP or handed off externally, as otherwise you can find out something about the user. And from that perspective always performing a (COOP-like) top-level navigation would be best when it's translated to HTTP.

Anyway, I don't care too strongly about the exact name, just want to make sure we consider these aspects.

ArthurSonzogni commented:
That's fair enough! Thank you!

Then would you be happy with `allow-top-navigation-to-custom-protocols` name?

annevk commented:
Yeah that works for me.

mtrootyy commented on source:
```diff
@@ -31238,19 +31238,25 @@ interface <dfn interface>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
+   <li><code
+   data-x="attr-iframe-sandbox-allow-top-navigation-by-user-activation">allow-top-navigation-by-user-activation</code></li>
+   <li><code
+   data-x="attr-iframe-sandbox-allow-top-navigation-to-custom-protocols">allow-top-navigation-to-protocols</code></li>
```
Incorrect: `>allow-top-navigation-to-protocols</code>`
  Correct:  `>allow-top-navigation-to-custom-protocols</code>`

ArthurSonzogni replied:
Thanks!

ArthurSonzogni reviewed (commented):
I will implement the proposed spec in Chrome, assuming you don't have any comments.

annevk commented on source:
```diff
@@ -31238,19 +31238,25 @@ interface <dfn interface>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
-  and <code data-x="attr-iframe-sandbox-allow-downloads">allow-downloads</code>.</p>
+  tokens</span> that are <span>ASCII case-insensitive</span>. The allowed values are:</p>
+  <ul class="brief">
+   <li><code data-x="attr-iframe-sandbox-allow-downloads">allow-downloads</code></li> <li><code
```
Nit: missing newline.

annevk commented on source:
```diff
@@ -31238,19 +31238,25 @@ interface <dfn interface>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
-  data-x="attr-iframe-sandbox-allow-top-navigation-by-user-activation">allow-top-navigation-by-user-activation</code>,
-  and <code data-x="attr-iframe-sandbox-allow-downloads">allow-downloads</code>.</p>
+  tokens</span> that are <span>ASCII case-insensitive</span>. The allowed values are:</p>
+  <ul class="brief">
```
Nit: newline before `<ul>`.

annevk commented on source:
```diff
@@ -31271,9 +31280,9 @@ interface <dfn interface>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
-  keywords re-enable forms, modal dialogs, screen orientation lock, the pointer lock API, popups,
-  the presentation API, scripts, and the creation of unsandboxed <span data-x="auxiliary browsing
-  context">auxiliary browsing contexts</span> respectively. The <code
+  keywords re-enable forms, modal dialogs, screen orientation lock,
+  the pointer lock API, popups, the presentation API, scripts, and the creation of unsandboxed <span
+  data-x="auxiliary browsing context">auxiliary browsing contexts</span> respectively. The <code
```
Nit: these changes seem unnecessary.

annevk commented on source:
```diff
@@ -83825,6 +83841,13 @@ interface <dfn interface>BarProp</dfn> {
+   <dt>The <dfn export>sandboxed custom protocols navigation browsing context flag</dfn></dt>
+
+   <dd>
+    <p>This flag prevents navigations toward URLs unsupported by the user agent to be <span
+    data-x="hand-off to external software">handed off to external software</span>.</p>
```
Maybe we should rephrase this to just talk about non-fetch schemes or some such?

annevk commented on source:
```diff
@@ -83935,6 +83958,12 @@ interface <dfn interface>BarProp</dfn> {
      data-x="attr-iframe-sandbox-allow-downloads">allow-downloads</code></dfn> keyword.</p></li>
 
+     <li><p>The <span>sandboxed custom protocols navigation browsing context flag</span>, unless
+     <var> tokens</var> contains either the <dfn attr-value for="iframe/sandbox"><code
```
Nit: remove leading space inside `<var>`.

annevk commented on source:
```diff
@@ -88462,26 +88491,26 @@ interface <dfn interface>Location</dfn> { // but see also <a href="#the-location
-     data-x="attr-iframe-sandbox-allow-top-navigation-by-user-activation">allow-top-navigation-by-user-activation</code>
-     token and <var>hasTransientActivation</var> is true</p></li>
+    <p>Return without invoking the external sofware package if all of these conditions hold:</p>
+    <ul class="brief">
```
Nit: keep newline before `<ul>`.

annevk commented on source:
```diff
@@ -88462,26 +88491,26 @@ interface <dfn interface>Location</dfn> { // but see also <a href="#the-location
     <p class="note">Navigation inside an iframe toward external software can be seen by users as a
-    new popup or a new top-level navigation. Therefore, we enforcing sandbox flags
-    appropriately.</p>
+    new popup or a new top-level navigation. That's why its is allowed in sandboxed iframe only when
```
Nit: `sandboxed <code>iframe</code> elements` seems slightly cleaner.

annevk commented on source:
```diff
@@ -124458,16 +124487,25 @@ interface <dfn interface>External</dfn> {
-          "<code data-x="attr-iframe-sandbox-allow-same-origin">allow-same-origin</code>",
-          "<code data-x="attr-iframe-sandbox-allow-scripts">allow-scripts</code>" and
-          "<code data-x="attr-iframe-sandbox-allow-top-navigation">allow-top-navigation</code>"
+      <ul class="brief">
+       <li><code data-x="attr-iframe-sandbox-allow-downloads">allow-downloads</code></li> <li><code
+       data-x="attr-iframe-sandbox-allow-forms">allow-forms</code></li>
+       <li><code data-x="attr-iframe-sandbox-allow-modals">allow-modals</code></li>
+       <li><code
+       data-x="attr-iframe-sandbox-allow-orientation-lock">allow-orientation-lock</code></li>
+       <li><code data-x="attr-iframe-sandbox-allow-pointer-lock">allow-pointer-lock</code></li>
+       <li><code data-x="attr-iframe-sandbox-allow-popups">allow-popups</code></li>
+       <li><code
+       data-x="attr-iframe-sandbox-allow-popups-to-escape-sandbox">allow-popups-to-escape-sandbox</code></li>
+       <li><code data-x="attr-iframe-sandbox-allow-presentation">allow-presentation</code></li>
+       <li><code data-x="attr-iframe-sandbox-allow-same-origin">allow-same-origin</code></li>
+       <li><code data-x="attr-iframe-sandbox-allow-scripts">allow-scripts</code></li>
+       <li><code data-x="attr-iframe-sandbox-allow-top-navigation">allow-top-navigation</code></li>
+       <li><code
+       data-x="attr-iframe-sandbox-allow-top-navigation-by-user-activation">allow-top-navigation-by-user-activation</code></li>
+       <li><code
+       data-x="attr-iframe-sandbox-allow-top-navigation-to-custom-protocols">allow-top-navigation-to-custom-protocols</code></li>
+      </ul>
```
I think we should keep the double quotes around the values here as it was originally. That's more consistent with the rest of the table. It's probably okay to turn this into a list though.

annevk reviewed (commented):
I had a look through the actual proposed text and found a number of small things to address. Overall this seems good though and we might pick this up soonish in Firefox.

ArthurSonzogni commented:
Thanks!
